### DISCUSSION TITLE
Separator degrees of freedom

### DISCUSSION THREAD

**Discussion by juanbaquero4** [IDAES/idaes-pse#1067] [Q&A]
Hi y'll, 

Hope you can help me with this issue.  I am trying to run a water injection process in order to solve a MILP optimization to improve energy efficiency.

I am trying to add a separator unit to split the water stream in two directions, however, after coding the parameters,  the number of degrees of freedom is 4, and after adding the arcs is - 5, and don't understand how to overcome this issue. Also , is neccesary to define the split_fraction and I am having an issue with this. There is the code:
```
m.fs.S25 = Separator (default={"property_package": m.fs.properties,
                                "outlet_list": [m.fs.Inj_4 , m.fs.PV_25], 
                                "num_outlets": 2 })
```

Where:
```
from idaes.generic_models.unit_models.product import Product
m.fs.Inj_4 = Product (default={"property_package": m.fs.properties})
f_pad = 100
m.fs.Inj_4.inlet.flow_mol[0].fix(f_pad) 
```
and:
```
m.fs.PV_25 = Valve (default={"property_package": m.fs.properties})
fin = 202 # mol/s # 🤷‍♂️ how to fis this value as the difference from the product to Water_PAD from the total transfered from the pumps or from the CV of the valve
pin = 1895058 # Pa # 🤷‍♂️ how to fix this with a pressure from from pumps 24 (arcs?)
pout = 1007058 # Pa # 🤷‍♂️ how to set this from the CV of the valve 
tin = 340 # K
hin = iapws95.htpx(T=tin*units.K, P=pin*units.Pa) # J/mol
#* Calculate the flow coefficient to give 1000 mol/s flow with given P
cv = fin/numpy.sqrt(pin - pout)/0.5
#* set inlet
m.fs.PV_25.inlet.enth_mol[0].fix(hin)
m.fs.PV_25.inlet.flow_mol[0].fix(fin)
m.fs.PV_25.inlet.flow_mol[0].unfix()
m.fs.PV_25.inlet.pressure[0].fix(pin)
m.fs.PV_25.outlet.pressure[0].fix(pout)
m.fs.PV_25.Cv.fix(cv)
m.fs.PV_25.valve_opening.fix(0.3)
```
and:
```
m.fs.pa = Arc(source=m.fs.pump_24_A.outlet, destination=m.fs.S25.inlet)
m.fs.pb = Arc(source=m.fs.pump_24_B.outlet, destination=m.fs.S25.inlet)
m.fs.pc= Arc(source=m.fs.pump_24_C.outlet, destination=m.fs.S25.inlet)
```

**Comment by andrewlee94** [comment]
Hi @juanbaquero4, I can see a number of problems to start with. Before I being however, I will note that you appear to be using IDAES v1.13 - as of February we will be releasing IDAES v2.0 and the v1 series will no longer be supported. We strongly encourage you to make the transition as soon as possible.

1. Firstly, when you are creating the arcs, you have three different outlets (from pumps) going to the inlet of a single separator. You can only have a single Arc connected to each Port, otherwise the problem will be over-specified. From what I can see, I think you need to have a Mixer in there to combine the streams from all the pumps first.
2. Next, you have a number of issues when you define the separator. Firstly, you only need to define one of `num_outlets` and `outlet_list` - `num_outlets` creates the specified number of outlets with default names, whilst `outlet_list` lets you provide a list of custom names. In the case where you provide both, we make sure they are consistent, and then just use `outlet_list`. More importantly however, `outlet_list` should be a list of names for the outlets, not a list of things you want to connect to - you need to create the Arcs later to do the connections. In this case, I *think* that the names will be converted to strings and used as the names for the outlets, but there could be unexpected behaviour from this.
3. For the degrees of freedom, I would need to see the full flowsheet, but one thing you did not show in your code above was fixing the split fraction for the Separator. You need something like `m.fs.S25.split_frac['to_PV_25'].fix(0.5)` - i.e. specifying that 50% of the flow goes to the outlet named `'to_PV_25'`.
4. Next, I think you are a bit confused about how to define the flowsheet connectivity. From the definition of the separator, you are saying that one stream enters and then gets split between the value (`PV25`) and some unit `Inj_4`. However, the next thing you do is to fix the inlet states of `PV25` - normally you would fix the inlet to the separator (or whatever is further upstream) and then solve for the inlet state of the valve. That is not to say you can't do this (equation oriented modeling lets you do things like this), but you need to know what you are doing and this doesn't match the general description of the process you gave.

**Reply by juanbaquero4** [reply]
Hi @andrewlee94, 

First of all, thank you for your answer, I upgraded the idaes version to 2.0.0a0

1.  As you suggested, I created a Mixer and I defined the "inlet_list"
2.  As you suggested, I fixed again the outlet_list with the new streams
3.  For degrees of freedom, I linked the property of unit models downstream/upstream, trying to keep a consistent material/energy balance, but at the end of the exercise remain  14 degrees of freedom to define
4. Kindly, I attach the link to the notebook where you can find the sketch of the promp and the code. (I am coding only the part within the red dash square)

https://github.com/juanbaquero4/PILE-OPTIMIZATION.git

Is it possible to arrange a small metting to clarify some concerns? I would really like to take the best advantage of this powerful tool. 

Regards:

**Reply by andrewlee94** [reply]
@juanbaquero4 Have you gone through the [IDAES Tutorials](https://github.com/IDAES/examples-pse/tree/main/src/Tutorials/Basics) yet, especially the HDA flowsheet example? This shows you how to construct a flowsheet for an example process, and includes mixers and splitters. From what I can see in you example, you seem to have a number of misunderstandings about how things work, and these tutorials should help show you how things should be done.

The first thing I see is when you create the Mixer, you are setting `"inlet_list":[m.fs.pump_24_A.inlet.flow_mol[0], m.fs.pump_24_B.inlet.flow_mol[0], m.fs.pump_24_C.inlet.flow_mol[0]]`. The inlet list should be a list of names (as strings), not a list of things you want to connect to (e.g. `"inlet_list": ["pump_24_A", "pump_24_B", "pump_24_C"]`). Also, I note that you are using the molar flow rate in each of those streams when you are creating the inlet list, so I am not sure what you are trying to achieve.

Following from this, I see you only create one Arc (between the Mixer outlet and the Separator inlet). This means that the pumps are not connected to the Mixer. You seem to be under the impression that by providing the unit names in the Mixer inlet list that this creates a connection between the units - this is **not** the case. The `inlet_list` just provides the names that will be used for the inlets, and you need to make connections later using Arcs. The same applies for the outlets of a Separator.

Next, whilst this is not a causing the issues you are seeing, you should be aware that having three pumps going into a mixer like that can cause unexpected behaviour with regards to pressure. Note that the default Mixer is a pressure minimization unit; i.e., the outlet pressure is equal to the lowest of the inlet pressures. Effectively, this means that the pressure of the other two streams has no effect on the outlet.

The next issue I see is that you are fixing the inlet conditions to the valve, even though you intend it to be connected to the outlet of the Separator. If all the connections had been made correctly, this would mean the system would be over-specified as the outlet of the separator would be defined by the mass and energy balances upstream, thus you could not then fix the conditions entering the valve.

**Reply by juanbaquero4** [reply]
Thank you @andrewlee94  for your insights! I finally managed to converge the system. Now I'm am upgrading it before change it to dynamic model,  and I really appreciate your hand in this; I am trying to add pump curves performance and I have the following questions:

1. Is possible to add pump curvs to a unit,model Pump? I tried and a message pop out saying:  _Scalar Pump do not have attribute of performance curve_ 
2. In case I have to change the unit model from pump to compressor. Does the compressor have any issues performing `m.fs.properties = iapws95.Iapws95ParameterBlock(default={"phase_presentation":PhaseType.L})` ?  I tried to run with unit model `compressor` but it doesn't  converge 
3. For the power performance of the unit model. Do I have  add it as a constraint? I didn't see it in the sub-block `perfomance curve`
4. Is neccesary to add the add the lines at the end of the code `assert` ? 
5. Do you have an example how to change a model from static to dynamic? 
6. Kindly I attache the code adding performance curves to unit model, I appreciate any comments regarding this:

```
from pyomo.environ import ConcreteModel, SolverFactory, units, value
from idaes.core import FlowsheetBlock
from idaes.generic_models.unit_models.pressure_changer import Pump
from idaes.generic_models.properties import iapws95
import pytest

solver = SolverFactory('ipopt')
m = ConcreteModel()
m.fs = FlowsheetBlock(default={"dynamic": False})
m.fs.properties = iapws95.Iapws95ParameterBlock()
m.fs.unit = Pump (default={
    "property_package": m.fs.properties,
    "support_isentropic_performance_curves":True})

Q = 302
# Add performance curves
@m.fs.unit.performance_curve.Constraint(m.fs.config.time)
def pc_isen_eff_eqn(b, t):
    # main pressure changer block parent of performance_curve
    prnt = b.parent_block()
    return prnt.efficiency_isentropic[t] == (-0.00000000513*Q**4)+(0.00000168*Q**3)-(0.000243*Q**2)+(0.0208*Q)-(0.0000155)
@m.fs.unit.performance_curve.Constraint(m.fs.config.time)
def pc_isen_head_eqn(b, t):
    # divide both sides by 1000 for scaling
    return b.head_isentropic[t]/1000 == ((-0.00000203*Q**4)+(0.000562*Q**3)-(0.0382*Q**2)-(1.76*Q)+991)*2984/1000*units.J/units.kg

# set inputs
m.fs.unit.inlet.flow_mol[0].fix(Q)  # mol/s
Tin = 350  # K
Pin = 101325  # Pa
hin = iapws95.htpx(Tin*units.K, Pin*units.Pa)
m.fs.unit.inlet.enth_mol[0].fix(hin)
m.fs.unit.inlet.pressure[0].fix(Pin)

m.fs.unit.initialize()
solver.solve(m, tee=False)

# assert value(m.fs.unit.efficiency_isentropic[0]) == pytest.approx(0.9, rel=1e-3)
# assert value(m.fs.unit.deltaP[0]) == pytest.approx(-3e5, rel=1e-3)

```

Regards;

Juan

**Reply by andrewlee94** [reply]
1. Yes, you can add performance curves, but they generally need to be added as new Constraints (much like you are trying to do later). I think there are a few more advanced models that have pre-built hooks for performance curves, but those will identify the hooks in the configuration arguments.
2. Compressors can handle phase transitions, but they use a more complex equation for the fluid work (isentropic performance versus pressure times flowrate). As such, they will always be harder to converge, especially when phase transitions occur. However, we do regularly use turbines (which are basically the same model) for steam cycles which do involve some phase transition.
2b. However, I should note that the single phase version of the IAPWS95 model is generally less stable than the mixed phase formulation. It is often better to use the mixed phase form, and do a post check of the results to ensure you didn't cross the phase boundary.
3. It depends on the model you are using, but if you are using the ones from `idaes/models/unit_models` then you need to add a `Constraint`.
4. The `assert` statements are only there for us to test the models - you do not need them. They are there so that we can automatically run the code and get errors raise if the result does not match the expected value.
5. We do not have any good example on how to do this - there are some dynamic flowsheets but they are probably too complex to serve as examples. In general, there are two ways to solve dynamic models in IDAES.
5a. Using an EO solver with full discretization of the time domain. This is best for optimization of dynamic systems, but results in a big model and can be hard to solve.

    i) Set the `Flowsheet` to be dynamic and provide units and range for time - this will automatically make all unit models dynamic unless you override it at the unit level, which adds the accumulation terms for you.
    ii) Apply the Pyomo DAE tools to expand the time domain and apply a numerical discretization - see the Pyomo documentation for this. You want a forward finite difference or collocation method for this. You will also need to carefully choose the number of finite elements you use.
    iii) Next, you need to add some form of pressure-flow constraints to your model. These are case specific, so you will need to work out what these are yourself.
    v) Set the necessary initial conditions for the model - there is a tool that will automatically set the initial time derivatives to zero if you want to start from steady state.

5b. Using an integrator, such as PETSc. We have some examples of that in our documentation that you can look at. This approach is often more robust and easier to solve, but is far less useful for optimization.

Finally, some comments on the code:

1. You are setting the `Flowsheet` to be steady-state - this means the model will not be dynamic.
1. You need to use `m.fs.time` for the time domain (or `unit.flowsheet().time`) - config blocks are only used for passing in initial configurations and should never be used in constraints.
2. Why is `Q` a constant in the performance constraint? You should be using the flowrate in the actual unit (`prnt.control_volume.properties_out[t].flow_vol`).
3. Whilst it is OK for starting out, hard coding scaling will make things harder for you if you ever need to use this model for different conditions. IDAES has some scaling tools to help with this (and we plan to improve them this year).
4. Your constraints are not unit consistent - this is fine as long as you don;t want to test for unit consistency, but as soon as you do they will fail. The coefficients need to be given units for this to work.

**Reply by juanbaquero4** [reply]
Thank you again! @andrewlee94 .. 

I was trying to follow the HDA example to set costraints for `m.fs.unit.outlet.pressure` and `m.fs.unit.efficiency_pump` and `m.fs.unit.efficiency_pump` , however, the values calculated in the constraints are not the same as calculated as when you print the values (showed in the code):

I set `Q` for a fixed curve performance which would be later proposed as (22, 32), which I haven't set it yet. After I have model a pump properly I will  fix a MILP with the minimum energy consumption of the system with parallel pumps... Thank you again for your help..

At the end I will try to run a dynamic model...

```

Q = 22 # mol/s 
P_disc_constraint = Constraint (expr = m.fs.unit.outlet.pressure[0] == 
                     ((-3.1795*Q**4)+(150.43*Q**3)-(2258.8*Q**2)*(22975*Q)+3000000))  # *units Pa 
eff_constraint = Constraint (expr = m.fs.unit.efficiency_pump ==  
                   (0.00002*Q**3)-(0.0022*Q**2)+(0.0774*Q)+(0.009)) 

print (value(m.fs.unit.outlet.pressure[0]))
100000.0  

print  ((-3.1795*Q**4)+(150.43*Q**3)-(2258.8*Q**2)-(22975*Q)+3000000)
2172193.728  # real value 
```

**Reply by andrewlee94** [reply]
when you printed those values above, had you successfully converged the model yet? If the solver had successfully converged then those values should be the same.

**Reply by juanbaquero4** [reply]
Yes, I code it in separate lines and at the end the solver converged, but the value that is taken  in the solver for ` m.fs.unit.outlet.pressure[0]` is 100000.0   instead of 2172193.728 , that sould be the value calculated with Q = 22 mol/s and I dont know the reason.

This curves are adjusted at 60 hz.

**Reply by andrewlee94** [reply]
I would need to see your script (and the solver output) to say more, but when did you print the values (before or after the solve)?

**Reply by juanbaquero4** [reply]
https://github.com/juanbaquero4/pump-example.git

before the solver. And te report of shows another value for discharge pressure (5E08)

**Reply by andrewlee94** [reply]
That is why the values are different - creating a constraint does not automatically change the value of a variable. you have to call the solver fist to find a solution to the problem, after which the values should be the same.

**Reply by juanbaquero4** [reply]
But, How can I initialize the pump with 2 degrees of freedom? or:
-  should I put a (x) value for `m.fs.unit.outlet.pressure[0]`
- run the solver 
- add constrint
- run the solver again?

**Reply by andrewlee94** [reply]
You should never initialize a model with degrees of freedom, as you can never be sure what the solution will be (if one is even found). You should always start with a well defined problem, and then open up degrees of freedom later.

**Reply by juanbaquero4** [reply]
Thank you! I get your point... let me work on it

**Reply by juanbaquero4** [reply]
hi @andrewlee94, thank you again for your insights! I finallly managed optimize a model of parallel pumps with a fix frequency!
Now, I am trying to add a variable of frequency which is linked with the calculation of pump's constraints, but the model do not initialize with this new modification. 
I would really appreciate your help in how to overcome this. Here is the code https://github.com/juanbaquero4/pump-frequency.git
Thank you again for your help!

**Reply by andrewlee94** [reply]
Without looking at the code, my first suggestion is to initialize the models with fixed frequency, and then add the variable frequency constraints/changes. Most of our initialization routines start with a simplified model (constraints deactivated and variables fixed in their place) that we solve, and then progressively add detail back into the model and resolving at each step. Also, I would advise only adding the variable frequency to one unit at a time - it makes it much easier to debug when you have only one change that could have caused an issue,

**Reply by juanbaquero4** [reply]
hi @andrewlee94 , I initialized the model with fixed variables,  then I unfixed variables and set constraints for ` unit.inlet.flow_mol[t]` that is calculated from the intial flow ` unit.inlet.flow_mol[0]`  and the variable `unit.frequency` , after that I set the constraints  `unit.outlet.pressure[t]` and `unit.bhp[t]` that are calculated  with the variable `unit.frequency` and ` unit.inlet.flow_mol[0]` at each frequency with the performance curve at @60 hz.   But it doesn't work.  
I was also thinking to make a loop for each constraint but it doen't work either. I really appreciate your advices.
Here is the code https://github.com/juanbaquero4/pump-frequency.git
Thank you again!

**Reply by andrewlee94** [reply]
I am going through the notebook you uploaded and making some comments as I go:

1. You are using an older version of IDAES (v1 series). We just released v2 which makes some performance improvements but also has some API changes, so you might want to look at making the change.
2. When indexing Constraints you should use `unit.flowsheet().time` or `m.fs.time` - do not use the config block as that is not guaranteed to be the actual time set.
3. Do not reinitialize the model after adding the new constraints - that just throws away all the work you had already done, and will have issues as now there are additional constraints the unit does not know about. Just call a solver again and hope.
4. You are trying to change too much in one go - notably you should separate adding the performance curves from adding the objective and have a solve in between. You want to add detail gradually, and have a square problem at each step up until you are ready to do the optimization. I started by leaving the inlet flow fixed and commenting out the objective.
5. When setting units you should be using Pyomo unit objects, not strings. `from pyomo.environ import units; units=units.Pa`.
6. I am not sure why you are adding a separate equation for the brake horsepower - the model already has a calculation for work.
7. Making the above changes, when I call IPOPT it reports an infeasible solution, which more often than not implies that there is a bound violation. Sure enough, removing the bounds on efficiency allowed the model to solve. The efficiency was coming out to be 0.88273, hence the solver failure.

Here is a notebook with my edits. Note that I am using IDAES v2 so I had to make the API changes I mentioned, so you will probably need to revert those (they were all in the first cell).
[jb_notebook_edit.txt](https://github.com/IDAES/idaes-pse/files/10995103/jb_notebook_edit.txt)

**Reply by juanbaquero4** [reply]
@andrewlee94  thank you so much, now it is working well ! I also unfix the `m.fs.unit.inlet.flow_mol[0].unfix()` and add a constraint for a minimum outlet pressure and it works well!! 
Thank you for your suggestions and I already fix the other recommendations. 
`unit.bhp[t]`it is just to compare from the value calculated based upon the performance curve.

**Reply by juanbaquero4** [reply]
Hi @andrewlee94 ! Which solver is supported by IDAES for MINLP.  I was reading the documentation but is not clear for me what additional solvers apart from `Ipopt` and `PETSC ` am I able to use. 
Thank you again!

**Reply by andrewlee94** [reply]
IDAES works through Pyomo, so any solver supported by Pyomo should work. See here for more information: https://pyomo.readthedocs.io/en/stable/solving_pyomo_models.html#supported-solvers.

**Reply by juanbaquero4** [reply]
I am trying to solve with this code recommended for MINLP:
```

opt = SolverFactory('mindtpy').solve(m,
                                     strategy='FP',
                                     mip_solver='glpk',
                                     nlp_solver='ipopt',
                                     tee=True) 
```

however, this message error pop out " Could not locate the 'glpsol' executable, which is required for solver 'glpk'" . 
I was trying to overcome this issue installing `%pip install glpk ` in jupyter notebook but doesn't wrok and later  through conda environment with the command `conda install -c conda-forge glpk` and doesn't work either.

**Reply by andrewlee94** [reply]
That looks like a solver installation issue, which unfortunately I cannot help with. You would need to check the `glpk` documentation for more information on that.

**Reply by juanbaquero4** [reply]
Yes, I was trying to figure out hoyw to run it but is was possible only though google colab. 
I was trying to run the model with the solver `'couenne'`  which is installed by defeault, but it doesn't run and the following error pops out : "Solver (asl) did not exit normally"
I ran the the same model with the solver `'bonmin'` and it gave me an optimal solution and I wa strying to compare it with this other solver. 
Can you guide why is it happening? Thank you!

**Reply by andrewlee94** [reply]
In short, no. Unfortunately I know very little about the MINLP solvers.

However, you should be aware that every solver is different and it is not unusual to have cases where one solver will find an optimal solution where as the other does not, and there can be many reasons for this. Each solver has a different approach to solving the problem, and sometimes one path to the solution is feasible whilst the other is not. It is also possible to have degenerate problems (problems with infinite solutions) where one solver will be satisfied by any one of those solutions whilst another might recognize that it is in a degenerate space and fail (IDAES has some prototype tools for identifying degenerate problems in `idaes/core/util/model_diagnostics`, but note that these are WIP).

Finally - one note on the error message you reported. The specific message indicates a failure in the solver interface, not the solver itself. This might just be due to a less than helpful error message, but it could also indicate a failure in the interface itself. That however I cannot help with.

**Reply by juanbaquero4** [reply]
Thank you @andrewlee94  again! I am almost meeting the objective. 

I am trying to set a conditional as a constraint, however, this error message pop up. How is possible to configure a conditional including a variable?

```
def eq_Pdesc_4A (P_4_A, P_4_B):
    if m.fs.P_4_A.inlet.flow_mol[0] > 10:
        return  m.fs.P_4_A.outlet.pressure[0] == m.fs.P_4_B.outlet.pressure[0]
    else:
        return Constraint.Skip
    
m.fs.eq_Pdesc_4A = Constraint(rule=eq_Pdesc_4A)
```

PyomoException: Cannot convert non-constant Pyomo expression (10  <  fs.P_4_A.control_volume.properties_in[0.0].flow_mol) to bool.
This error is usually caused by using a Var, unit, or mutable Param in a
Boolean context such as an "if" statement, or when checking container
membership or equality. For example,
    >>> m.x = Var()
    >>> if m.x >= 1:
    ...     pass
and
    >>> m.y = Var()
    >>> if m.y in [m.x, m.y]:
    ...     pass
would both cause this exception.

**Reply by andrewlee94** [reply]
@juanbaquero4 There are a number of issues with what you are tying to do there.

The first is that rules are only used when building the constraint, and are not actually seen by the solver. I.e. your `if` statement is only evaluated once when the constraint is constructed, and the constraint will end up having only the expression that was true at that time. In order to have conditional constraints, you either need to use `expr_if` statements in the rule (which makes the IF part of the expression), or use a numerical equivalent (e.g. a complementarity constraint).

The next issue is understanding the limitations of conditional constraints - the general requirement is that a conditional constraint should be continuous at least (i.e. that at the switch point both sides of the IF provide the same result) and that the first derivative should be monotonic across the switch (i.e. you don;t want to have the first derivative go from positive to negative). The constraint you are trying to write violates both of those, as one side applied a Constraint and the other does not - this means that the degrees of freedom in the problem will change as you cross the switch. This seems like a very bad idea.

Finally, there is the issue that gives rise to the exception you are seeing. The issue there is the statement in the IF itself: `m.fs.P_4_A.inlet.flow_mol[0] > 10`. When evaluating a Pyomo component in a boolean check, you need to use the `value` of the component, not the component itself; i.e. `value(m.fs.P_4_A.inlet.flow_mol[0]) > 10`. This is because Pyomo components create Pyomo `expressions` instead of evaluating, and an `expression` cannot be part of the IF.

**Reply by juanbaquero4** [reply]
@andrewlee94 hi,  thank you, again!

Now I am using DAE to model a recycle  in a water transfer system. After running the model, seems that the arcs from the pump `(P_4)`  to a pipeline `(L_4)` are not connected, the pressure is totally different from one unit to another, I do not understand what is the reason that this is happening.  The control valve to recycle `(PV_4) ` open to 100 % to control the pressure, but no exists any fluctuation of pressure  in the unit  `(L_4)`

Also, the water inlet to the pump goes from 12 mol/s (initial condition) to a value >1500 mol/s in the firsts seconds, I set several steps to reduce the value, but the peak continues appearing in the same time step.  (I reckon it is possible that I do not use the recall to calculate variables from the pump, I tried as it is posted on IDAES documnetation but it does not run)

Besides those point, Do exists an IDAES tool to report the cumulative `unit.inflow_mol` calculated through DAE

Here is the file https://github.com/juanbaquero4/pump-example.git

Imnensily grateful for your recommendations.

**Reply by andrewlee94** [reply]
@juanbaquero4 The first problem is see with your example notebook is that the solver reported an infeasible solution - thee is no point at looking at any results beyond that point as the solver has told you it has not converged. This means that it is possible for there to be mismatches in the results.

you will need to dig into your model and work out why it is reporting an infeasible solutions. The first thing to do with an infeasible solution is to check for any bound violations in the model - this is the most common cause of infeasible solutions. You should go though the full model and look for variables that are approaching bounds and work out why, and then either revise the operating conditions to avoid the violation or relax the bound. However, you should always ask yourself why the bound was there before relaxing it - in property packages for example bounds often indicate the range over which the model correlations were fitted and thus represent the "trust region" of the model - relaxing those bounds mean you are now extrapolating with all the risks associated with that.

The second thing you should do is run the final solve with `tee=True` and inspect the output from IPOPT. There is a lot of information in there that can be useful for suggesting what might be causing the solver to fail.

**Reply by juanbaquero4** [reply]
Hi @andrewlee94  thank you! I made some changes in the model however the mistake persists in t=1.  I changed the values for initial conditions and cheking the report, the initial conditions corresponds correct values. Thus, I am thinking the error from the calculation in t=1 comes from `unit.model pump` calculations, so I coded the constratint to calculate discharge pressure and efficiency (used in previous models), however, this error pops out:

`InitializationError: fs.P_4 failed to initialize successfully. Please check the output logs for more information.`
-> `m, solver = create_model(time_set=[0, 5], nfe=10, tee=False)`
-> `m.fs.P_4.initialize()`
->`flags = blk.initialize_build(*args, **kwargs)`
->`super().initialize_build(state_args=state_args, outlvl=outlvl, solver=solver, optarg=optarg` 

The constraints were built as follows: 

```
    @m.fs.P_4.Constraint(m.fs.time)
    def Pdesc_calculation(P_4, t):
        return (
            (P_4.outlet.pressure[t]) == (
            (-3874*P_4.control_volume.properties_out[t].flow_mol**2)+
            (-37153*P_4.control_volume.properties_out[t].flow_mol)+
            (2860479))
        )
    
    @m.fs.P_4.Constraint(m.fs.time)
    def eff_calculation(P_4, t):
        return (
            (P_4.efficiency_pump[t]) == (
            (-0.0068*P_4.control_volume.properties_out[t].flow_mol**2)+
            (0.1483*P_4.control_volume.properties_out[t].flow_mol)+
            (0.00252))
        )
    
```

I tried to built a function to recall the constraints as it is made from valve, however as it is documented for` unit.model Pressure Changer`, the `isentropic.performance.curve` is allowed to `turbine` not to `pumps. `

here is the code: https://github.com/juanbaquero4/pump-example.git

As always,  I inmensily appreciate your help.

**Reply by andrewlee94** [reply]
@juanbaquero4 Before you rush off to change the model, you need to look at why the initialization failed; there are many different reasons the model might have failed to solve and each have a different fix. You should start by looking at the final solver status, as well as the number of variables and constraints in the problem and then look over the solver output logs to see what they are telling you.

**Reply by juanbaquero4** [reply]
@andrewlee94  As you suggested I took a review of all the variables in the model, I tried to change the ways to calculate to variables for the unit  `pump` and I noticed the following:

- In `t= 0 `with `nef=1` , the solver found a optimal solution and the variables were correct in all the units

- in` t = 1 `with `nef =1`,  the solver finished due the # of iterations, I checked the results and the `pump.outlet.pressure` displayed in the graph does not correspond to the value that should be calculated from function` (Pdesc)`. Instead to rise it declines, even with the recycle `PV_4 `close.  (t=1 the value of flow is 11 so the pressure should rise)

- ¿how is possible to check the values as report, I've tried with the fucntion m.fs.report (1), but displayed the same result for m.fs.report(0), also I tried to print the values for instance in this case the flow `value(m.fs.P_4.control_volume.properties_out[t].flow_mol), ` but does not work. 

-  If` nfe` is changed to a value different to 1, the values of flow_pump and `inlet_flow.pump` inmediatly rises from a value of 10 to 5000, the same for `pump.outlet.pressure `do not understand the reason why those values are taken for the `unit_pump.`

I am really struglling trying to understand. I would really appreciate your help, I've been deeping on the model but I couldn't find a way to make it converge. 

here is the code: https://github.com/juanbaquero4/pump-example.git

As always thank you!

**Reply by andrewlee94** [reply]
There is not much point in trying to inspect the model results if the final solve was not successful - any result you do get back will be representative of the failed state at best. It can be useful for debugging infeasible models, but not for cases of max iterations. What you need to be looking at is the solver logs and see why it is failing.

From a quick look at the logs for the final solve I see a few signs of your problem. First of all, I note that you have two degrees of freedom in the model (80 variables but only 78 constraints); this is almost definitely not what you want as it mans the model is under-defined (you can do optimization on dynamic models, but you need to get the deterministic model working first). In these cases, there often is no answer to your problem hence the maximum iterations. Secondly, I see a lot of steps with regularization (lg(rg), should be "-"), small alpha values (should be 1) and high line search values (ls, should be 1). This tends to suggest that the model does not have a well defined feasible point and that the partial derivatives in the area around the current iteration are very small - this would be the case for an under-defined problem.

You might want to start with a smaller flowsheet to start with - I would suggest starting with just the valve and controller with a defined inlet condition. Once you understand how to get a single unit model working properly, you can then start working backward.

One other thing that just came to mind, is that you might need to think about the behaviour of the INJ unit. You may need some sort of pressure-flow constraint here too as otherwise the flow in htis stream might be undefined.

**Comment by Robbybp** [comment]
Edited to put the code in blocks with fixed-width text, FYI.
